test/e2e/network_segmentation.go

```diff
@@ -0,0 +1,425 @@
+							cs,
+							f.Namespace.Name,
+							serverPod.GetName(),
+							namespacedName(serverPod.Namespace, netConfig.name),
+						)
+						Expect(err).NotTo(HaveOccurred())
+						const netPrefixLengthPerNode = 24
+						By(fmt.Sprintf("asserting the server pod IP %v is from the configured range %v/%v", serverIP, cidr, netPrefixLengthPerNode))
+						subnet, err := getNetCIDRSubnet(cidr)
+						Expect(err).NotTo(HaveOccurred())
+						Expect(inRange(subnet, serverIP)).To(Succeed())
+					}
+
+					By("asserting the *client* pod can contact the server pod exposed endpoint")
+					Eventually(func() error {
+						return reachToServerPodFromClient(cs, serverPodConfig, clientPodConfig, serverIP, port)
+					}, 2*time.Minute, 6*time.Second).Should(Succeed())
+				}
+			},
+			Entry(
+				"two pods connected over an IPv4 network",
+				networkAttachmentConfigParams{
+					name:           userDefinedNetworkName,
+					networkName:    userDefinedNetworkName,
+					topology:       "layer2",
+					cidr:           userDefinedNetworkIPv4Subnet,
+					excludeCIDRs:   []string{externalServiceIPv4IP + "/32"},
+					primaryNetwork: true,
+				},
+				*podConfig(
+					"client-pod",
+					withNodeSelector(map[string]string{nodeHostnameKey: workerOneNodeName}),
+				),
+				*podConfig(
+					"server-pod",
+					withCommand(func() []string {
+						return httpServerContainerCmd(port)
+					}),
+					withNodeSelector(map[string]string{nodeHostnameKey: workerTwoNodeName}),
+				),
+			),
+		)
+
+		//DescribeTable(
+		//	"allows egressing from the pods to the outside network",
+		//	func(netConfigParams networkAttachmentConfigParams, clientPodConfig podConfiguration, serverPodConfig podConfiguration) {
+		//
+		//	},
+		//)
+	})
+})
+
+type podOption func(*podConfiguration)
+
+func podConfig(podName string, opts ...podOption) *podConfiguration {
+	pod := &podConfiguration{
+		name: podName,
+	}
+	for _, opt := range opts {
+		opt(pod)
+	}
+	return pod
+}
+
+func withCommand(cmdGenerationFn func() []string) podOption {
+	return func(pod *podConfiguration) {
+		pod.containerCmd = cmdGenerationFn()
+	}
+}
+
+func withNodeSelector(nodeSelector map[string]string) podOption {
+	return func(pod *podConfiguration) {
+		pod.nodeSelector = nodeSelector
+	}
+}
+
+func podIPForUserDefinedPrimaryNetwork(k8sClient clientset.Interface, podNamespace string, podName string, attachmentName string) (string, error) {
+	pod, err := k8sClient.CoreV1().Pods(podNamespace).Get(context.Background(), podName, metav1.GetOptions{})
+	if err != nil {
+		return "", err
+	}
+	netStatus, err := userDefinedNetworkStatus(pod, attachmentName)
+	if err != nil {
+		return "", err
+	}
+
+	if len(netStatus.IPs) == 0 {
+		return "", fmt.Errorf("attachment for network %q without IPs", attachmentName)
+	}
+	if len(netStatus.IPs) > 1 {
+		return "", fmt.Errorf("attachment for network %q with more than one IP", attachmentName)
+	}
+	return netStatus.IPs[0].IP.String(), nil
```

Comment:
for dualstack, we'll need to parametrize this. Check what we're doing on secondary nets, and do that. Should just work.